test `flow_table_0` rerun on April 14 on GitLab CI

For the record `flow_table_0` had a rerun last night, although it passed in the next attempt. 

It's worth analyzing this later on.


```
rerun: 0
tests/test_e2e_20_flow_manager.py::TestE2EFlowManager::test_070_flow_table_0: 2023-04-14,06:01:41.190601 - 2023-04-14,06:01:55.931930
self = <tests.test_e2e_20_flow_manager.TestE2EFlowManager object at 0x7feaee1a9ca0>
    def test_070_flow_table_0(self):
>       self.flow_table_0()
tests/test_e2e_20_flow_manager.py:930: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <tests.test_e2e_20_flow_manager.TestE2EFlowManager object at 0x7feaee1a9ca0>
restart_kytos = False
    def flow_table_0(self, restart_kytos=False):
        """Tests if, after adding a flow in another
        table outside kytos, the flow is removed."""
    
        s1 = self.net.net.get('s1')
        s1.dpctl('add-flow', 'table=0,in_port=1,actions=output:2')
    
        if restart_kytos:
            # restart controller keeping configuration
            self.net.start_controller(enable_all=True)
            self.net.wait_switches_connect()
        else:
            self.net.reconnect_switches()
    
        time.sleep(10)
    
        s1 = self.net.net.get('s1')
        flows_s1 = s1.dpctl('dump-flows')
>       assert len(flows_s1.split('\r\n ')) == BASIC_FLOWS, flows_s1
E       AssertionError:  cookie=0xac00000000000001, duration=25.521s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:03 actions=CONTROLLER:65535
E          cookie=0xac00000000000001, duration=25.474s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:02 actions=CONTROLLER:65535
E          cookie=0x0, duration=14.715s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1" actions=output:"s1-eth2"
E          cookie=0xab00000000000001, duration=13.703s, table=0, n_packets=17, n_bytes=714, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
E         
E       assert 4 == 3
E        +  where 4 = len([' cookie=0xac00000000000001, duration=25.521s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:...n=13.703s, table=0, n_packets=17, n_bytes=714, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'])
E        +    where [' cookie=0xac00000000000001, duration=25.521s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:...n=13.703s, table=0, n_packets=17, n_bytes=714, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'] = <built-in method split of str object at 0x7feadc013910>('\r\n ')
E        +      where <built-in method split of str object at 0x7feadc013910> = ' cookie=0xac00000000000001, duration=25.521s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:0...on=13.703s, table=0, n_packets=17, n_bytes=714, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'.split
tests/test_e2e_20_flow_manager.py:927: AssertionError
=========================== rerun test summary info ============================
RERUN tests/test_e2e_20_flow_manager.py::TestE2EFlowManager::test_070_flow_table_0
= 205 passed, 6 skipped, 11 xfailed, 5 xpassed, 799 warnings, 1 rerun in 11317.01s (3:08:37) =
```

--- a/tests/test_e2e_11_mef_eline.py
+++ b/tests/test_e2e_11_mef_eline.py
@@ -417,6 +417,8 @@ class TestE2EMefEline:
         response = requests.patch(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
         assert response.status_code == 400, response.text
 
+        time.sleep(5)
+
         # Check for VLAN 999
         api_url = f'{KYTOS_API}/topology/v3/interfaces/tag_ranges'
         response = requests.get(api_url)

--- a/tests/test_e2e_15_mef_eline.py
+++ b/tests/test_e2e_15_mef_eline.py
@@ -241,6 +241,8 @@ class TestE2EMefEline:
         assert response.status_code == 201, response.text
         assert 'circuit_id' in response.json()
 
+        time.sleep(5)
+
         # Verify if EVC tag has been allocated
         topo_url = KYTOS_API + "/topology/v3/interfaces/tag_ranges"
         response = requests.get(topo_url)
@@ -345,6 +347,8 @@ class TestE2EMefEline:
         response = requests.post(api_url, json=payload)
         assert response.status_code == 201, response.text
 
+        time.sleep(5)
+
         intf_id = '00:00:00:00:00:00:00:01:1'
         api_url = KYTOS_API + f'/topology/v3/interfaces/{intf_id}/tag_ranges'
         response = requests.get(api_url)
@@ -444,6 +448,8 @@ class TestE2EMefEline:
         response = requests.post(api_url, json=payload)
         assert response.status_code == 201, response.text
         circuit_id = response.json()["circuit_id"]
+
+        time.sleep(5)
 
         expected = [[1, 9], [16, 3798], [3800, 4094]]
         intf_id = '00:00:00:00:00:00:00:01:1'

--- a/tests/test_e2e_20_flow_manager.py
+++ b/tests/test_e2e_20_flow_manager.py
@@ -1,5 +1,6 @@
 import json
 import time
+import re
 
 import requests
 
@@ -77,7 +78,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         # Make sure that the flow that was sent is on /v2/stored_flows
         dpid = "00:00:00:00:00:00:00:01"
@@ -138,7 +142,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         sw = self.net.net.get("s1")
         flows_sw = sw.dpctl("dump-flows")
@@ -255,7 +262,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         for sw_name in ['s1', 's2', 's3']:
             sw = self.net.net.get(sw_name)
@@ -350,7 +360,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         stored_flows = f'{KYTOS_API}/flow_manager/v2/stored_flows/?dpids={switch_id}&cookie_range=1&cookie_range=3'
         response = requests.get(stored_flows)
@@ -405,7 +418,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')
@@ -457,7 +473,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         # Make sure that flows are soft deleted on /v2/stored_flows
         response = requests.get(
@@ -481,13 +500,13 @@ class TestE2EFlowManager:
         for i in range(1, 4):
             dpid = f"00:00:00:00:00:00:00:0{i}"
             assert dpid in data
-            assert len(data[dpid]["flows"]) == BASIC_FLOWS, data[dpid]
+            assert len(data[dpid]["flows"]) == BASIC_FLOWS, f"dpid={dpid} data={data[dpid]}"
 
         for sw_name in ['s1', 's2', 's3']:
             sw = self.net.net.get(sw_name)
             flows_sw = sw.dpctl('dump-flows')
             assert len(flows_sw.splitlines()) == BASIC_FLOWS, flows_sw
-            assert 'actions=output:2' not in flows_sw
+            assert 'actions=output:2' not in flows_sw, flows_sw
 
     def test_026_delete_flows_cookie_mask_range(self):
         """Test deleting flows with cookie range mask and persistence."""""
@@ -557,7 +576,10 @@ class TestE2EFlowManager:
         self.net.start_controller(enable_all=True, del_flows=True)
         self.net.wait_switches_connect()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        # specially because we used "del_flows=True"
+        time.sleep(15)
 
         # Make sure that flows are soft deleted on /v2/stored_flows
         response = requests.get(
@@ -831,6 +853,7 @@ class TestE2EFlowManager:
         payload = {
             "flows": [
                 {
+                    "cookie": 0x99,
                     "priority": 10,
                     "idle_timeout": 360,
                     "hard_timeout": 1200,
@@ -858,13 +881,15 @@ class TestE2EFlowManager:
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')
         assert len(flows_s1.splitlines()) == BASIC_FLOWS + 1, flows_s1
-        assert 'in_port=1' in flows_s1
+        assert len(re.findall("cookie=0x99.*in_port=1 .*output:2", flows_s1)) == 1, flows_s1
 
         # Modify the actions and verify its modification
-        s1.dpctl('mod-flows', 'actions=output:3')
+        # we use output:7 which is a port not in use, to avoid further
+        # problems with loop detection and mismatch links
+        s1.dpctl('mod-flows', 'actions=output:7')
         flows_s1 = s1.dpctl('dump-flows')
         assert 'actions=output:2' not in flows_s1
-        assert 'actions=output:3' in flows_s1
+        assert 'actions=output:7' in flows_s1
 
         if restart_kytos:
             # restart controller keeping configuration
@@ -879,8 +904,8 @@ class TestE2EFlowManager:
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')
         assert len(flows_s1.splitlines()) == BASIC_FLOWS + 1, flows_s1
-        assert 'actions=output:3' not in flows_s1
-        assert 'in_port=1' in flows_s1
+        assert 'actions=output:7' not in flows_s1
+        assert len(re.findall("cookie=0x99.*in_port=1 .*output:2", flows_s1)) == 1, flows_s1
 
     def test_040_replace_action_flow(self):
         self.replace_action_flow()
@@ -1068,6 +1093,8 @@ class TestE2EFlowManager:
             for flow in sw_flows[BASIC_FLOWS: BASIC_FLOWS + length]:
                 assert flow["state"] in {"installed", "pending"}
 
+        time.sleep(5)
+
         s1, s2, s3 = self.net.net.get('s1', 's2', 's3')
         flows_s1 = s1.dpctl('dump-flows')
         flows_s2 = s2.dpctl('dump-flows')
@@ -1107,6 +1134,8 @@ class TestE2EFlowManager:
             for flow in sw_flows[BASIC_FLOWS: BASIC_FLOWS + length]:
                 assert flow["state"] == "deleted"
 
+        time.sleep(5)
+
         flows_s1 = s1.dpctl('dump-flows')
         flows_s2 = s2.dpctl('dump-flows')
         flows_s3 = s3.dpctl('dump-flows')
@@ -1129,6 +1158,8 @@ class TestE2EFlowManager:
                       headers={'Content-type': 'application/json'})
         assert response.status_code == 202, response.text
 
+        time.sleep(5)
+
         s1, s2, s3 = self.net.net.get('s1', 's2', 's3')
         flows_s1 = s1.dpctl('dump-flows')
         flows_s2 = s2.dpctl('dump-flows')
@@ -1145,6 +1176,9 @@ class TestE2EFlowManager:
         response = requests.delete(api_url, data=json.dumps(payload),
                       headers={'Content-type': 'application/json'})
         assert response.status_code == 202, response.text
+
+        time.sleep(5)
+
         flows_s1 = s1.dpctl('dump-flows')
         flows_s2 = s2.dpctl('dump-flows')
         flows_s3 = s3.dpctl('dump-flows')
@@ -1168,6 +1202,8 @@ class TestE2EFlowManager:
         response = requests.post(api_url, data=json.dumps(payload),
                                  headers={'Content-type': 'application/json'})
         assert response.status_code == 202, response.text
+
+        time.sleep(5)
 
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')
@@ -1230,6 +1266,8 @@ class TestE2EFlowManager:
             for flow in sw_flows[BASIC_FLOWS: BASIC_FLOWS + length]:
                 assert flow["state"] in {"installed", "pending"}
 
+        time.sleep(5)
+
         s1, s2, s3 = self.net.net.get('s1', 's2', 's3')
         flows_s1 = s1.dpctl('dump-flows')
         flows_s2 = s2.dpctl('dump-flows')
@@ -1268,6 +1306,8 @@ class TestE2EFlowManager:
         for sw_flows, length in test_list:
             for flow in sw_flows[BASIC_FLOWS: BASIC_FLOWS + length]:
                 assert flow["state"] == "deleted", flow
+
+        time.sleep(5)
 
         flows_s1 = s1.dpctl('dump-flows')
         flows_s2 = s2.dpctl('dump-flows')

--- a/tests/test_e2e_20_flow_manager.py
+++ b/tests/test_e2e_20_flow_manager.py
@@ -835,7 +835,9 @@ class TestE2EFlowManager:
         else:
             self.net.reconnect_switches()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        time.sleep(15)
 
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')
@@ -898,7 +900,9 @@ class TestE2EFlowManager:
         else:
             self.net.reconnect_switches()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        time.sleep(15)
 
         # Check that the flow keeps the original setting
         s1 = self.net.net.get('s1')
@@ -953,7 +957,9 @@ class TestE2EFlowManager:
         else:
             self.net.reconnect_switches()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        time.sleep(15)
 
         flows_s1 = s1.dpctl('dump-flows')
         assert len(flows_s1.splitlines()) == BASIC_FLOWS + 1, flows_s1
@@ -979,7 +985,9 @@ class TestE2EFlowManager:
         else:
             self.net.reconnect_switches()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        time.sleep(15)
 
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')
@@ -1005,7 +1013,9 @@ class TestE2EFlowManager:
         else:
             self.net.reconnect_switches()
 
-        time.sleep(10)
+        # STATS_INTERVAL is set to 7 seconds, we should wait at least
+        # two cycles to account for possible Overlapping requests
+        time.sleep(15)
 
         s1 = self.net.net.get('s1')
         flows_s1 = s1.dpctl('dump-flows')

